### PR TITLE
feat: add type_generic and is_dttm to table metadata

### PR DESCRIPTION
**Pull request by zhaoyongjie**
### SUMMARY

For the current the Superset, `is_temporal` is used to determine if a column is a DateTime column instead of the column type (Only the columns marked with `is_temporal` can be used for time picker). Superset doesn't mark a time column as `is_temporal` when creating a new dataset or syncing an existing dataset.

This PR introduce `is_dttm` and `general_column` fields to the dataset metadata. so that it can be mark a DateTime column with `is_temporal`.

### update existing columns
https://user-images.githubusercontent.com/2016594/119802345-0ee26780-bf11-11eb-8b37-31e3d2d0e689.mp4

### add new columns

https://user-images.githubusercontent.com/2016594/119802789-713b6800-bf11-11eb-9d41-0bf86c51049c.mp4

### ADDITIONAL INFORMATION

- [x] Has associated issue: fixes https://github.com/apache/superset/issues/14754
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: .pylintrc [hunk 1]
```diff
@@ -81,7 +81,7 @@ confidence=
...
+disable=long-builtin,dict-view-method,intern-builtin,suppressed-message,no-absolute-import,unpacking-in-except,apply-builtin,delslice-method,indexing-exception,old-raise-syntax,print-statement,cmp-builtin,reduce-builtin,useless-suppression,coerce-method,input-builtin,cmp-method,raw_input-builtin,nonzero-method,backtick,basestring-builtin,setslice-method,reload-builtin,oct-method,map-builtin-not-iterating,execfile-builtin,old-octal-literal,zip-builtin-not-iterating,buffer-builtin,getslice-method,metaclass-assignment,xrange-builtin,long-suffix,round-builtin,range-builtin-not-iterating,next-method-called,parameter-unpacking,unicode-builtin,unichr-builtin,import-star-module-level,raising-string,filter-builtin-not-iterating,using-cmp-argument,coerce-builtin,file-builtin,old-division,hex-method,missing-docstring,too-many-lines,ungrouped-imports,import-outside-toplevel,raise-missing-from,super-with-arguments,bad-option-value,too-few-public-methods,too-many-locals
```
Comment (zhaoyongjie) [inline .pylintrc:84]: disable `too-many-locals` lint rule. There are too many ignore in the codebase for this rule.

File: superset/connectors/sqla/models.py [hunk 2]
```diff
@@ -677,13 +677,25 @@ def external_metadata(self) -> List[Dict[str, str]]:
...
                     if isinstance(col["type"], TypeEngine):
-                        col["type"] = db_engine_spec.column_datatype_to_string(
+                        db_type = db_engine_spec.column_datatype_to_string(
                             col["type"], db_dialect
                         )
+                        type_spec = db_engine_spec.get_column_spec(db_type)
+                        col.update(
+                            {
+                                "type": db_type,
+                                "generic_type": type_spec.generic_type
```
Comment (villebro) [inline superset/connectors/sqla/models.py:687]: I believe this is called `type_generic` on the data model: https://github.com/apache/superset/blob/b6f00e69e156098e9e1569d11f0f7831bac1024f/superset/connectors/base/models.py#L568-L578

**Review by villebro — COMMENTED** [review]
Small comment about naming. I also wonder if we should start deprecating the use of `is_dttm` and start promoting the use of `type_generic`?

File: superset/connectors/sqla/models.py (hunk 2 above)
Comment (zhaoyongjie) [reply superset/connectors/sqla/models.py]: good catch, rename it. btw, the method `external_metadata(self)` is for getting metadata from the table in database(In other words, not in superset saved metadata).

**Comment by zhaoyongjie** [comment]
> Small comment about naming. I also wonder if we should start deprecating the use of `is_dttm` and start promoting the use of `type_generic`?

I think in the explore page should return type_generic(derived from type) instead of is_dttm or type. but in the sqllab should return database original column type.

File: superset-frontend/src/datasource/DatasourceEditor.jsx [hunk 3]
```diff
@@ -414,13 +415,15 @@ class DatasourceEditor extends React.PureComponent {
           type: col.type,
           groupby: true,
           filterable: true,
+          is_dttm: col.is_dttm,
         });
         results.added.push(col.name);
-      } else if (currentCol.type !== col.type) {
+      } else if (currentCol.type !== col.type || col.is_dttm) {
```
Comment (villebro) [inline superset-frontend/src/datasource/DatasourceEditor.jsx:421]: Should we make this more specific?
```js
      } else if (currentCol.type !== col.type || currentCol.is_dtm !== col.is_dttm) {
```

Comment (zhaoyongjie) [reply superset-frontend/src/datasource/DatasourceEditor.jsx:421]: good catch!

**Review by villebro — APPROVED** [review]
LGTM

**Pull request merged**